<?xml version="1.0"?>

<launch>
    <arg name="open_rviz" default="true" />
    <arg name="enable_laser_scan_matcher" default="false" />
    <arg name="enable_rf2o_laser_odometry" default="false" />

    <!-- File ist broken: Unconnected tf trees because of the LaserScanMatcher.
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/2017-11-13-16-22-33.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/2017-11-15-13-44-05.bag" />

    Bei den folgenden Aufnahmen gilt:
    - Das Topic "beacon" ist vom Typ "mrpt_msgs/ObservationRangeBeacon". Korrigierte Entfernung.
    - Das Topic "beacon_raw" ist vom Typ "mrpt_msgs/ObservationRangeBeacon"
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-24-50.bag" /> Kein Tag vorhanden
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-31-28.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-33-03.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-34-10.bag" /> Nur zwei Tag
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-36-25.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-12-41-35.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-13-25-18.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-13-25-38.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-13-28-16.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-13-31-05.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-11-17-13-33-13.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-12-18-11-50-57.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-12-18-11-53-54.bag" />
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-12-18-11-53-54_filtered.bag" />

    Bei den folgenden Aufnahmen gilt:
    - Das Topic "beacon" ist vom Typ "mrpt_msgs/ObservationRangeBeacon"
    - Das Topic "beacon_raw" ist vom Typ "ro_slam_with_uwb/Beacon"
    -->
    <arg name="rosbag_bagfile" value="$(find ro_slam_with_uwb)/bags/Record_2017-12-18-11-53-54.bag" />
    <arg name="rosbag_args" value="--clock --delay=0 --rate=2.0 --start=0" />

    <!-- Parameter -->
    <param name="/use_sim_time" value="true" />
    

    <!-- Nodes -->
    <!-- -->
    <node
        name="rosbag_node" pkg="rosbag" type="play"
        output="log"
        args="$(arg rosbag_bagfile) $(arg rosbag_args)" />
    <!-- -->

    <!--
        http://wiki.ros.org/hector_trajectory_server?distro=kinetic
        https://answers.ros.org/question/209224/show-robot-trajectory-in-rviz-real-time/
		https://github.com/PickNikRobotics/rviz_visual_tools
		https://answers.ros.org/question/209580/odometry-to-path-rviz/
    -->
    <node 
        pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server"
        output="screen">

        <param name="target_frame_name" value="odom"/>
        <param name="source_frame_name" value="base_link" />
        <param name="trajectory_update_rate" value="10.0" />
        <param name="trajectory_publish_rate" value="10"/>
    </node>
    <!-- -->

    <!--
        http://wiki.ros.org/hector_mapping?distro=kinetic
        http://wiki.ros.org/hector_slam/Tutorials/SettingUpForYourRobot
    -->
    <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">
        <param name="scan_topic" value="scan" />
        <param name="base_frame" value="base_link" />
        <param name="odom_frame" value="odom" />
        <param name="map_frame" value="map" />

        <param name="pub_map_odom_transform" value="true" />

        <param name="map_resolution" value="0.02" />
        <param name="laser_min_dist" value="0.05" />
        <param name="laser_max_dist" value="30.0" />
    </node>
    <!-- -->

    <!--
    <node
        pkg="ro_slam_with_uwb" type="particlecloud_to_csv.py" name="particlecloud_to_csv_node"
        output="screen">

        <param name="csv_file_name" value="$(arg rosbag_bagfile)" />
    </node>
    -->

    <!--
    <node
        pkg="ro_slam_with_uwb" type="beacon_to_csv.py" name="beacon_to_csv_node"
        output="screen">

        <param name="csv_file_name" value="$(arg rosbag_bagfile)" />
        <param name="beacon_topic_name" value="beacon_raw" />
    </node>
    -->

    <!--
        http://wiki.ros.org/rf2o_laser_odometry
    -->
    <node
        if="$(arg enable_rf2o_laser_odometry)"
        pkg="rf2o_laser_odometry" type="rf2o_laser_odometry_node" name="rf2o_laser_odometry"
        output="screen">

        <param name="laser_scan_topic" value="scan"/>
        <param name="odom_topic" value="odom" />
        <param name="publish_tf" value="true" />
        <param name="base_frame_id" value="base_link"/>
        <param name="odom_frame_id" value="odom" />
        <param name="init_pose_from_topic" value="" />
        <param name="freq" value="15.0"/>
        <param name="verbose" value="false" />
    </node>

    <!--
        http://wiki.ros.org/laser_scan_matcher
    -->
    <node
        if="$(arg enable_laser_scan_matcher)"
        pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node"
        output="screen">

        <param name="base_frame" value="base_link"/>
        <param name="fixed_frame" value="odom" />
        <param name="use_odom" value="true" />
        <param name="use_imu" value="false" />
    </node>
    <!-- -->

    <!--
        http://wiki.ros.org/gmapping
    -->
    <!--
    <node
        pkg="gmapping" type="slam_gmapping" name="slam_gmapping"
        output="screen">

        <remap from="base_frame" to="base_link"/>
        <param name="map_frame" value="map"/>
        <param name="odom_frame" value="odom"/>
        <param name="map_update_interval" value="1"/>
        <param name="delta" value="0.01" />

        <param name="xmin" value="5" />
        <param name="ymin" value="5" />
        <param name="xmax" value="5" />
        <param name="ymax" value="5" />

        <param name="maxRange" value="25" />
        <param name="maxUrange" value="25" />
    </node>
    -->

    <!-- Include -->
    <!--
    <include file="$(find ro_slam_with_uwb)/launch/mrpt_rbpf_slam.launch" />
    -->
    <!-- -->
    <include file="$(find ro_slam_with_uwb)/launch/robotino_state_publisher.launch" />
    <!-- -->

    <!-- -->
    <group if="$(arg open_rviz)">
        <include file="$(find ro_slam_with_uwb)/launch/rviz.launch" />
    </group>
    <!-- -->
</launch>